### ISSUE TITLE
[feature req] scatter search targets of threads to optimize search across multiple devices

### ISSUE THREAD

**Issue by LamaRaz** [sharkdp/fd#559]
when searching across multiple devices, cpu utilization is rarely the bottleneck but rather io is.

an option to scatter the search targets of threads would speed up search massively when searching in /media/ for example across multiple external harddrives.

**Comment by sharkdp** [comment]
> when searching across multiple devices, cpu utilization is rarely the bottleneck but rather io is.

Even when not searching across multiple devices, IO is the bottleneck. Even more so, if the disk cache is not warm (https://github.com/sharkdp/fd-benchmarks, https://github.com/sharkdp/hyperfine#warmup-runs-and-preparation-commands).

> an option to scatter the search targets of threads would speed up search massively 

Do you have any measurements that back up your claim or is this statement based on intuition? (I'm not implying that it isn't true)

`fd` already "scatters" search paths across multiple threads. You can even control the number of threads with `-j`/`--threads`. If you limit to one thread, you should see a significant decline in search performance. If you have a lot of devices you want to search, it might be useful to increase the number of threads to the number of devices. I have never tried that. The default number of threads is equal to the number of cpu cores.

Perhaps surprisingly, using multiple threads is even beneficial if the search tree is rooted on a single disk. This is due to IO scheduling (and caching) effects. If you search multiple directories/disks, there will not be a unique mapping from threads to devices/folders, but I that is not necessary from a performance point of view.

To summarize: I'm not sure what your feature request is about? What else would you like `fd` to do?

**Comment by LamaRaz** [comment]
launching fd in the root directories of multiple devices finished much faster than launching fd in the directory containing all those devices with the number of threads equal (j2 for 4 devices in parallel and j8 for the latter case).

suggesting  that the threads didn't disperse themselves across the devices.

i don't have concrete timings

**Comment by sharkdp** [comment]
> launching fd in the root directories of multiple devices finished much faster than launching fd in the directory containing all those devices with the number of threads equal (j2 for 4 devices in parallel and j8 for the latter case).

I'm not sure I can follow. If you can tell me exactly which `fd` commands you executed, I might be able to look into this.

> i don't have concrete timings

What's the purpose of this ticket then?

Benchmarking of tools like `fd` is not trivial. You need to be aware of caching effects and explicitly perform either warm-cache or cold-cache benchmarks. See https://github.com/sharkdp/hyperfine#warmup-runs-and-preparation-commands and github.com/sharkdp/fd-benchmarks for more information.

> edit: if the dispersion is random, the solution may be to just spam more threads. is it?

I'm not even sure what the problem is?

**Comment by LamaRaz** [comment]
The commands was basic, fd -H -I -a -tf > a

There were reboots in between the tests and one was at least twice as fast as the other which is why going through the effort of timing was not considered important.

> I'm not even sure what the problem is?

If the dispersion is random, the problem is that 8 threads is too few to give a decent probability of randomly covering 4 (much less more) devices. The documentation does not say that in those instances you should use more threads, making one question if it's random after all?

**Comment by sharkdp** [comment]
> The documentation does not say that in those instances you should use more threads,

Because we don't even have concrete evidence for that.

I'm not going to make any changes to `fd` based on anecdotal results.

**Comment by LamaRaz** [comment]
I solved the problem with a script running fd in parallel but fd really does have a problem here

```
cd /media/
fd -H -I -a -tf -j 128 | fzf
```
/media/ contains 4 devices and first thing i noticed was that fzf was extremely slow, 2 of my devices are crap.

I ran

`dstat -d --disk-util --disk-tps`

surprise surprise, most of the time only one device was being utilized by all 128 threads, sometimes 2 and very rarely 3. never observed 4 and most of the time was spent with just 1.


so there is no good random dispersion of search threads, anyone running into this problem has to write a script and cannot rely solely on fd

**Comment by sharkdp** [comment]
Thank you.

> /media/ contains 4 devices and first thing i noticed was that fzf was extremely slow, 2 of my devices are crap.

Ok. Might be interesting to compare the read speeds of these devices via
```bash
sudo hdparm -t -T  /dev/sda
```
But I'm actually not sure if that really correlates with the filesystem-traversal-speed.

> surprise surprise, most of the time only one device was being utilized by all 128 threads, sometimes 2 and very rarely 3. never observed 4 and most of the time was spent with just 1.

But.. do all of these devices contain the same amount of files/folders? In the end of the search, I would always expect the threads to converge onto the largest and/or slowest device, because the search on the other devices is already finished.

> so there is no good random dispersion of search threads, anyone running into this problem has to write a script and cannot rely solely on fd

Again. I'm not saying you are wrong. But before we jump to conclusions I really want to see an actual benchmark that compares your two approaches (run `fd` once for all devices, run `fd` multiple times but restricted to a single device).

If the second approach is indeed faster, I completely agree that we should think about possible solutions to this problem.

**Comment by LamaRaz** [comment]
Unfortunately I don't have the time to spend to do a scientific study. I controlled enough variables to convince myself there is a problem here and seeing that I was able to work around the problem it no longer concerns me.

I would just like to point out that the number of files per device don't matter because there are enough to keep fd busy for a while and the disparity was observed within the first few minutes, long before the files were exhausted on any device.

Whether or not your suggested convergence eventually happens doesn't affect the initial observed inefficiency.

I also don't see the relevance of comparing speeds of the devices, the problem would still be present if all the speeds were identical. I merely highlighted the fact to illustrate that all fd threads got stuck on one crappy device, therefore completely ignoring to a long time the fast devices much less the other crappy one. I would suggest that this is highly undesirable behavior, but it's your project to lead.

**Comment by sharkdp** [comment]
> Unfortunately I don't have the time to spend to do a scientific study.

I'm sorry - I was not implying that you had to. I base `fd` design decisions on benchmark results. That does not mean that you have to perform these benchmarks.

> I controlled enough variables to convince myself there is a problem here and seeing that I was able to work around the problem it no longer concerns me.

Fair enough. I am grateful for your bug report. But please don't expect anyone to be super motivated to work on a ticket where someone claims that *"fd really does have a problem here"* without properly backing up that claim while being seemingly uninterested in getting the ticket closed.

> I also don't see the relevance of comparing speeds of the devices, the problem would still be present if all the speeds were identical.

I doubt that. With the current behavior of `fd`, I would actually *expect* most of the threads to be working on folders on the slower devices. Jobs (i.e. newly discovered folders) are distributed in a uniform way across all threads. However, if jobs on device A typically take time T_A and jobs on device B typically take time T_B, we actually expect to see a fraction of T_A/(T_A+T_B) of the jobs to be focused on device A (in the phase where both devices still yield new folders).

To determine if a particular job-distribution strategy is optimal, we need to specify what to optimize for. Two possibilities might be:

* Total `fd` runtime
* Fastest way to see first results

So far, we have been optimizing for total `fd` runtime.
